bam_to_mate_hist.py

```diff
@@ -11,147 +11,291 @@
+        stat_dict ({str: str/float}): a dictionary of relevant data (stats, file paths) regarding the lib to be used in the report.
+        outfile_name (str): path to which the report shall be written.
+
+    '''
+    options = {
+        'page-size': 'A4',
+        'dpi': 350,
+        'margin-top': '0.75in',
+        'margin-right': '0.75in',
+        'margin-bottom': '0.75in',
+        'margin-left': '0.75in',
+        'encoding': "UTF-8",
+        'custom-header': [
+            ('Accept-Encoding', 'gzip')
+        ],
+        'no-outline': None
+    }
+
+    template_path = os.path.join(qc_repo_path, "collateral", "HiC_QC_report_template.md")
+    style_path = os.path.join(qc_repo_path, "collateral", "style.css")
+    if not os.path.exists(template_path):
+        UserWarning("can't find markdown template at {0}! skipping making template.".format(
+            qc_repo_path)
+        )
+        sys.exit()
+
+    with open(template_path) as file:
+        str = file.read()
+        sub_str = str.format(**stat_dict)  # splat the statistics and path into the markdown, render as html
+        # print sub_str
+        html = md.markdown(sub_str, extensions=['tables', 'nl2br'])
+
+        # write out just html
+        html_out = open(outfile_name + "_qc_report.html", 'w')
+        html_out.write(html)
+        html_out.close()
+
+        # print html
+        pdfkit.from_string(html, outfile_name + "_qc_report.pdf", options=options, css=style_path)
+
+
+def extract_stats(stat_list, bamfile_handle, outfile_name, count_diff_refname_stub=False):
+    '''Make a dict of results and data suitable to be passed to the report template for splatting.
+
+    Args:
+        stat_list ([str]): the four statistics estimated from the bam file
+                           (stat_list = [diff_chr, dists, diff_stub, split_reads])
+        bamfile_handle (str): path to bam file
+        outfile_handle (str): path to where output files will be written
+        count_diff_refname_stub (bool): whether we are counting the contig name stub differences.
+
+    Returns:
+        stat_dict ({str:float/str}: mappings of the stats and data suitable to be consumed by report generator
+    '''
+    stat_dict = {}
+    stat_dict["BAM_FILE_PATH"] = os.path.split(bamfile_handle)[-1]
+    num_pairs = len(stat_list[1])
+    stat_dict["NUM_PAIRS"] = num_pairs
+
+    print "Histograms written to:", os.path.abspath(outfile_name + "_long.png"), os.path.abspath(outfile_name + "_short.png")
+    stat_dict["PATH_TO_LONG_HIST"] = os.path.abspath(outfile_name + "_long.png")
+    stat_dict["PATH_TO_SHORT_HIST"] = os.path.abspath(outfile_name + "_short.png")
+
+    print "Counts of zero distances (many is a sign of bad prep):"
+    unique, counts = np.unique(stat_list[1], return_counts=True)  # tabulates the distances, with indices as the dists
+    # print unique[-100:-1]
+    zero_dist = dict(zip(unique, counts))[0]  # first element is the zero-distances
```

Comment:
If you are going to make the dict, you might as well store it in a variable to use again unless you never need that dict again?

Reply:
we only use it this once.

Reply:
ok